File: examples/python/data_structures_comprehensive.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
        
    def append(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node
    
    def prepend(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node
    
    def delete(self, data):
        if not self.head:
            return
        if self.head.data == data:
            self.head = self.head.next
            return
        current = self.head
        while current.next:
            if current.next.data == data:
                current.next = current.next.next
                return
            current = current.next
    
    def find(self, data):
        current = self.head
        while current:
            if current.data == data:
                return True
            current = current.next
        return False
    
    def display(self):
        elements = []
        current = self.head
        while current:
            elements.append(current.data)
            current = current.next
        return elements
    
    def reverse(self):
        prev = None
        current = self.head
        while current:
            next_temp = current.next
            current.next = prev
            prev = current
            current = next_temp
        self.head = prev
    
    def get_middle(self):
        if not self.head:
            return None
        slow = self.head
        fast = self.head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow.data
    
    def has_cycle(self):
        if not self.head:
            return False
        slow = self.head
        fast = self.head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True
        return False

class Stack:
    def __init__(self):
        self.items = []
    
    def push(self, item):
        self.items.append(item)
    
    def pop(self):
        if not self.is_empty():
            return self.items.pop()
        return None
    
    def peek(self):
        if not self.is_empty():
            return self.items[-1]
        return None
    
    def is_empty(self):
        return len(self.items) == 0
    
    def size(self):
        return len(self.items)
    
    def display(self):
        return self.items.copy()

class Queue:
    def __init__(self):
        self.items = []
    
    def enqueue(self, item):
        self.items.append(item)
    
    def dequeue(self):
        if not self.is_empty():
            return self.items.pop(0)
        return None
    
    def front(self):
        if not self.is_empty():
            return self.items[0]
        return None
    
    def is_empty(self):
        return len(self.items) == 0
    
    def size(self):
        return len(self.items)
    
    def display(self):
        return self.items.copy()

class BinaryTreeNode:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class BinarySearchTree:
    def __init__(self):
        self.root = None
    
    def insert(self, data):
        if not self.root:
            self.root = BinaryTreeNode(data)
        else:
            self._insert_recursive(self.root, data)
    
    def _insert_recursive(self, node, data):
        if data < node.data:
            if node.left is None:
                node.left = BinaryTreeNode(data)
            else:
                self._insert_recursive(node.left, data)
        elif data > node.data:
            if node.right is None:
                node.right = BinaryTreeNode(data)
            else:
                self._insert_recursive(node.right, data)
    
    def search(self, data):
        return self._search_recursive(self.root, data)
    
    def _search_recursive(self, node, data):
        if node is None or node.data == data:
            return node is not None
        if data < node.data:
            return self._search_recursive(node.left, data)
        return self._search_recursive(node.right, data)
    
    def inorder_traversal(self):
        result = []
        self._inorder_recursive(self.root, result)
        return result
    
    def _inorder_recursive(self, node, result):
        if node:
            self._inorder_recursive(node.left, result)
            result.append(node.data)
            self._inorder_recursive(node.right, result)
    
    def preorder_traversal(self):
        result = []
        self._preorder_recursive(self.root, result)
        return result
    
    def _preorder_recursive(self, node, result):
        if node:
            result.append(node.data)
            self._preorder_recursive(node.left, result)
            self._preorder_recursive(node.right, result)
    
    def postorder_traversal(self):
        result = []
        self._postorder_recursive(self.root, result)
        return result
    
    def _postorder_recursive(self, node, result):
        if node:
            self._postorder_recursive(node.left, result)
            self._postorder_recursive(node.right, result)
            result.append(node.data)
    
    def find_min(self):
        if not self.root:
            return None
        current = self.root
        while current.left:
            current = current.left
        return current.data
    
    def find_max(self):
        if not self.root:
            return None
        current = self.root
        while current.right:
            current = current.right
        return current.data
    
    def height(self):
        return self._height_recursive(self.root)
    
    def _height_recursive(self, node):
        if not node:
            return -1
        left_height = self._height_recursive(node.left)
        right_height = self._height_recursive(node.right)
        return max(left_height, right_height) + 1

class Graph:
    def __init__(self):
        self.vertices = {}
    
    def add_vertex(self, vertex):
        if vertex not in self.vertices:
            self.vertices[vertex] = []
    
    def add_edge(self, vertex1, vertex2):
        if vertex1 not in self.vertices:
            self.add_vertex(vertex1)
        if vertex2 not in self.vertices:
            self.add_vertex(vertex2)
        
        self.vertices[vertex1].append(vertex2)
        self.vertices[vertex2].append(vertex1)
    
    def remove_edge(self, vertex1, vertex2):
        if vertex1 in self.vertices and vertex2 in self.vertices[vertex1]:
            self.vertices[vertex1].remove(vertex2)
        if vertex2 in self.vertices and vertex1 in self.vertices[vertex2]:
            self.vertices[vertex2].remove(vertex1)
    
    def get_vertices(self):
        return list(self.vertices.keys())
    
    def get_edges(self):
        edges = []
        for vertex in self.vertices:
            for neighbor in self.vertices[vertex]:
                if (neighbor, vertex) not in edges:
                    edges.append((vertex, neighbor))
        return edges
    
    def bfs(self, start_vertex):
        visited = set()
        queue = [start_vertex]
        result = []
        
        while queue:
            vertex = queue.pop(0)
            if vertex not in visited:
                visited.add(vertex)
                result.append(vertex)
                for neighbor in self.vertices.get(vertex, []):
                    if neighbor not in visited:
                        queue.append(neighbor)
        
        return result
    
    def dfs(self, start_vertex):
        visited = set()
        result = []
        
        def dfs_recursive(vertex):
            visited.add(vertex)
            result.append(vertex)
            for neighbor in self.vertices.get(vertex, []):
                if neighbor not in visited:
                    dfs_recursive(neighbor)
        
        dfs_recursive(start_vertex)
        return result

def test_data_structures():
    print("Testing LinkedList:")
    ll = LinkedList()
    ll.append(1)
    ll.append(2)
    ll.append(3)
    ll.prepend(0)
    print(f"List: {ll.display()}")
    print(f"Find 2: {ll.find(2)}")
    print(f"Middle element: {ll.get_middle()}")
    ll.reverse()
    print(f"Reversed: {ll.display()}")
    
    print("\nTesting Stack:")
    stack = Stack()
    stack.push(1)
    stack.push(2)
    stack.push(3)
    print(f"Stack: {stack.display()}")
    print(f"Pop: {stack.pop()}")
    print(f"Peek: {stack.peek()}")
    
    print("\nTesting Queue:")
    queue = Queue()
    queue.enqueue(1)
    queue.enqueue(2)
    queue.enqueue(3)
    print(f"Queue: {queue.display()}")
    print(f"Dequeue: {queue.dequeue()}")
    print(f"Front: {queue.front()}")
    
    print("\nTesting Binary Search Tree:")
    bst = BinarySearchTree()
    elements = [50, 30, 70, 20, 40, 60, 80]
    for elem in elements:
        bst.insert(elem)
    print(f"Inorder: {bst.inorder_traversal()}")
    print(f"Preorder: {bst.preorder_traversal()}")
    print(f"Postorder: {bst.postorder_traversal()}")
    print(f"Search 40: {bst.search(40)}")
    print(f"Min: {bst.find_min()}")
    print(f"Max: {bst.find_max()}")
    print(f"Height: {bst.height()}")
    
    print("\nTesting Graph:")
    graph = Graph()
    graph.add_edge('A', 'B')
    graph.add_edge('A', 'C')
    graph.add_edge('B', 'D')
    graph.add_edge('C', 'D')
    graph.add_edge('D', 'E')
    print(f"Vertices: {graph.get_vertices()}")
    print(f"Edges: {graph.get_edges()}")
    print(f"BFS from A: {graph.bfs('A')}")
    print(f"DFS from A: {graph.dfs('A')}")

if __name__ == "__main__":
    test_data_structures()

File: examples/go/data_structures_comprehensive.go
package main

import "fmt"

type Node struct {
	data int
	next *Node
}

type LinkedList struct {
	head *Node
}

func NewLinkedList() *LinkedList {
	return &LinkedList{}
}

func (ll *LinkedList) Append(data int) {
	newNode := &Node{data: data}
	if ll.head == nil {
		ll.head = newNode
		return
	}
	
	current := ll.head
	for current.next != nil {
		current = current.next
	}
	current.next = newNode
}

func (ll *LinkedList) Prepend(data int) {
	newNode := &Node{data: data, next: ll.head}
	ll.head = newNode
}

func (ll *LinkedList) Delete(data int) {
	if ll.head == nil {
		return
	}
	
	if ll.head.data == data {
		ll.head = ll.head.next
		return
	}
	
	current := ll.head
	for current.next != nil {
		if current.next.data == data {
			current.next = current.next.next
			return
		}
		current = current.next
	}
}

func (ll *LinkedList) Find(data int) bool {
	current := ll.head
	for current != nil {
		if current.data == data {
			return true
		}
		current = current.next
	}
	return false
}

func (ll *LinkedList) Display() []int {
	var elements []int
	current := ll.head
	for current != nil {
		elements = append(elements, current.data)
		current = current.next
	}
	return elements
}

func (ll *LinkedList) Reverse() {
	var prev *Node
	current := ll.head
	
	for current != nil {
		next := current.next
		current.next = prev
		prev = current
		current = next
	}
	
	ll.head = prev
}

func (ll *LinkedList) GetMiddle() int {
	if ll.head == nil {
		return -1
	}
	
	slow := ll.head
	fast := ll.head
	
	for fast != nil && fast.next != nil {
		slow = slow.next
		fast = fast.next.next
	}
	
	return slow.data
}

type Stack struct {
	items []int
}

func NewStack() *Stack {
	return &Stack{}
}

func (s *Stack) Push(item int) {
	s.items = append(s.items, item)
}

func (s *Stack) Pop() (int, bool) {
	if len(s.items) == 0 {
		return 0, false
	}
	
	index := len(s.items) - 1
	item := s.items[index]
	s.items = s.items[:index]
	return item, true
}

func (s *Stack) Peek() (int, bool) {
	if len(s.items) == 0 {
		return 0, false
	}
	return s.items[len(s.items)-1], true
}

func (s *Stack) IsEmpty() bool {
	return len(s.items) == 0
}

func (s *Stack) Size() int {
	return len(s.items)
}

func (s *Stack) Display() []int {
	result := make([]int, len(s.items))
	copy(result, s.items)
	return result
}

type Queue struct {
	items []int
}

func NewQueue() *Queue {
	return &Queue{}
}

func (q *Queue) Enqueue(item int) {
	q.items = append(q.items, item)
}

func (q *Queue) Dequeue() (int, bool) {
	if len(q.items) == 0 {
		return 0, false
	}
	
	item := q.items[0]
	q.items = q.items[1:]
	return item, true
}

func (q *Queue) Front() (int, bool) {
	if len(q.items) == 0 {
		return 0, false
	}
	return q.items[0], true
}

func (q *Queue) IsEmpty() bool {
	return len(q.items) == 0
}

func (q *Queue) Size() int {
	return len(q.items)
}

func (q *Queue) Display() []int {
	result := make([]int, len(q.items))
	copy(result, q.items)
	return result
}

type BinaryTreeNode struct {
	data  int
	left  *BinaryTreeNode
	right *BinaryTreeNode
}

type BinarySearchTree struct {
	root *BinaryTreeNode
}

func NewBinarySearchTree() *BinarySearchTree {
	return &BinarySearchTree{}
}

func (bst *BinarySearchTree) Insert(data int) {
	if bst.root == nil {
		bst.root = &BinaryTreeNode{data: data}
	} else {
		bst.insertRecursive(bst.root, data)
	}
}

func (bst *BinarySearchTree) insertRecursive(node *BinaryTreeNode, data int) {
	if data < node.data {
		if node.left == nil {
			node.left = &BinaryTreeNode{data: data}
		} else {
			bst.insertRecursive(node.left, data)
		}
	} else if data > node.data {
		if node.right == nil {
			node.right = &BinaryTreeNode{data: data}
		} else {
			bst.insertRecursive(node.right, data)
		}
	}
}

func (bst *BinarySearchTree) Search(data int) bool {
	return bst.searchRecursive(bst.root, data)
}

func (bst *BinarySearchTree) searchRecursive(node *BinaryTreeNode, data int) bool {
	if node == nil {
		return false
	}
	
	if node.data == data {
		return true
	}
	
	if data < node.data {
		return bst.searchRecursive(node.left, data)
	}
	
	return bst.searchRecursive(node.right, data)
}

func (bst *BinarySearchTree) InOrderTraversal() []int {
	var result []int
	bst.inOrderRecursive(bst.root, &result)
	return result
}

func (bst *BinarySearchTree) inOrderRecursive(node *BinaryTreeNode, result *[]int) {
	if node != nil {
		bst.inOrderRecursive(node.left, result)
		*result = append(*result, node.data)
		bst.inOrderRecursive(node.right, result)
	}
}

func (bst *BinarySearchTree) PreOrderTraversal() []int {
	var result []int
	bst.preOrderRecursive(bst.root, &result)
	return result
}

func (bst *BinarySearchTree) preOrderRecursive(node *BinaryTreeNode, result *[]int) {
	if node != nil {
		*result = append(*result, node.data)
		bst.preOrderRecursive(node.left, result)
		bst.preOrderRecursive(node.right, result)
	}
}

func (bst *BinarySearchTree) PostOrderTraversal() []int {
	var result []int
	bst.postOrderRecursive(bst.root, &result)
	return result
}

func (bst *BinarySearchTree) postOrderRecursive(node *BinaryTreeNode, result *[]int) {
	if node != nil {
		bst.postOrderRecursive(node.left, result)
		bst.postOrderRecursive(node.right, result)
		*result = append(*result, node.data)
	}
}

func (bst *BinarySearchTree) FindMin() (int, bool) {
	if bst.root == nil {
		return 0, false
	}
	
	current := bst.root
	for current.left != nil {
		current = current.left
	}
	
	return current.data, true
}

func (bst *BinarySearchTree) FindMax() (int, bool) {
	if bst.root == nil {
		return 0, false
	}
	
	current := bst.root
	for current.right != nil {
		current = current.right
	}
	
	return current.data, true
}

func (bst *BinarySearchTree) Height() int {
	return bst.heightRecursive(bst.root)
}

func (bst *BinarySearchTree) heightRecursive(node *BinaryTreeNode) int {
	if node == nil {
		return -1
	}
	
	leftHeight := bst.heightRecursive(node.left)
	rightHeight := bst.heightRecursive(node.right)
	
	if leftHeight > rightHeight {
		return leftHeight + 1
	}
	return rightHeight + 1
}

type Graph struct {
	vertices map[string][]string
}

func NewGraph() *Graph {
	return &Graph{vertices: make(map[string][]string)}
}

func (g *Graph) AddVertex(vertex string) {
	if _, exists := g.vertices[vertex]; !exists {
		g.vertices[vertex] = []string{}
	}
}

func (g *Graph) AddEdge(vertex1, vertex2 string) {
	g.AddVertex(vertex1)
	g.AddVertex(vertex2)
	
	g.vertices[vertex1] = append(g.vertices[vertex1], vertex2)
	g.vertices[vertex2] = append(g.vertices[vertex2], vertex1)
}

func (g *Graph) RemoveEdge(vertex1, vertex2 string) {
	if neighbors, exists := g.vertices[vertex1]; exists {
		g.vertices[vertex1] = remove(neighbors, vertex2)
	}
	
	if neighbors, exists := g.vertices[vertex2]; exists {
		g.vertices[vertex2] = remove(neighbors, vertex1)
	}
}

func remove(slice []string, item string) []string {
	for i, v := range slice {
		if v == item {
			return append(slice[:i], slice[i+1:]...)
		}
	}
	return slice
}

func (g *Graph) GetVertices() []string {
	var vertices []string
	for vertex := range g.vertices {
		vertices = append(vertices, vertex)
	}
	return vertices
}

func (g *Graph) GetEdges() [][]string {
	var edges [][]string
	visited := make(map[string]bool)
	
	for vertex := range g.vertices {
		for _, neighbor := range g.vertices[vertex] {
			edgeKey := vertex + "-" + neighbor
			reverseKey := neighbor + "-" + vertex
			
			if !visited[edgeKey] && !visited[reverseKey] {
				edges = append(edges, []string{vertex, neighbor})
				visited[edgeKey] = true
				visited[reverseKey] = true
			}
		}
	}
	
	return edges
}

func (g *Graph) BFS(startVertex string) []string {
	visited := make(map[string]bool)
	queue := []string{startVertex}
	var result []string
	
	for len(queue) > 0 {
		vertex := queue[0]
		queue = queue[1:]
		
		if !visited[vertex] {
			visited[vertex] = true
			result = append(result, vertex)
			
			for _, neighbor := range g.vertices[vertex] {
				if !visited[neighbor] {
					queue = append(queue, neighbor)
				}
			}
		}
	}
	
	return result
}

func (g *Graph) DFS(startVertex string) []string {
	visited := make(map[string]bool)
	var result []string
	
	g.dfsRecursive(startVertex, visited, &result)
	return result
}

func (g *Graph) dfsRecursive(vertex string, visited map[string]bool, result *[]string) {
	visited[vertex] = true
	*result = append(*result, vertex)
	
	for _, neighbor := range g.vertices[vertex] {
		if !visited[neighbor] {
			g.dfsRecursive(neighbor, visited, result)
		}
	}
}

func testDataStructures() {
	fmt.Println("Testing LinkedList:")
	ll := NewLinkedList()
	ll.Append(1)
	ll.Append(2)
	ll.Append(3)
	ll.Prepend(0)
	fmt.Printf("List: %v\n", ll.Display())
	fmt.Printf("Find 2: %t\n", ll.Find(2))
	fmt.Printf("Middle element: %d\n", ll.GetMiddle())
	ll.Reverse()
	fmt.Printf("Reversed: %v\n", ll.Display())
	
	fmt.Println("\nTesting Stack:")
	stack := NewStack()
	stack.Push(1)
	stack.Push(2)
	stack.Push(3)
	fmt.Printf("Stack: %v\n", stack.Display())
	if val, ok := stack.Pop(); ok {
		fmt.Printf("Pop: %d\n", val)
	}
	if val, ok := stack.Peek(); ok {
		fmt.Printf("Peek: %d\n", val)
	}
	
	fmt.Println("\nTesting Queue:")
	queue := NewQueue()
	queue.Enqueue(1)
	queue.Enqueue(2)
	queue.Enqueue(3)
	fmt.Printf("Queue: %v\n", queue.Display())
	if val, ok := queue.Dequeue(); ok {
		fmt.Printf("Dequeue: %d\n", val)
	}
	if val, ok := queue.Front(); ok {
		fmt.Printf("Front: %d\n", val)
	}
	
	fmt.Println("\nTesting Binary Search Tree:")
	bst := NewBinarySearchTree()
	elements := []int{50, 30, 70, 20, 40, 60, 80}
	for _, elem := range elements {
		bst.Insert(elem)
	}
	fmt.Printf("Inorder: %v\n", bst.InOrderTraversal())
	fmt.Printf("Preorder: %v\n", bst.PreOrderTraversal())
	fmt.Printf("Postorder: %v\n", bst.PostOrderTraversal())
	fmt.Printf("Search 40: %t\n", bst.Search(40))
	if min, ok := bst.FindMin(); ok {
		fmt.Printf("Min: %d\n", min)
	}
	if max, ok := bst.FindMax(); ok {
		fmt.Printf("Max: %d\n", max)
	}
	fmt.Printf("Height: %d\n", bst.Height())
	
	fmt.Println("\nTesting Graph:")
	graph := NewGraph()
	graph.AddEdge("A", "B")
	graph.AddEdge("A", "C")
	graph.AddEdge("B", "D")
	graph.AddEdge("C", "D")
	graph.AddEdge("D", "E")
	fmt.Printf("Vertices: %v\n", graph.GetVertices())
	fmt.Printf("Edges: %v\n", graph.GetEdges())
	fmt.Printf("BFS from A: %v\n", graph.BFS("A"))
	fmt.Printf("DFS from A: %v\n", graph.DFS("A"))
}

func main() {
	fmt.Println("=== Go Data Structures Demo ===")
	testDataStructures()
}